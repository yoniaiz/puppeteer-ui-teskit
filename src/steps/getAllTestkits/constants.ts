export const logs = {
  gettingTestkits: 'Getting testkits...',
  noTestkitsFound: 'No testkits found',
  notSupportedFileType: (filePath: string): string =>
    `File type not supported: ${filePath}`,
  testkitFileIsEmpty: (filePath: string): string =>
    `Teskit file is empty: ${filePath}`,
  testkitFileInvalidURL: (filePath: string, url: string): string =>
    `Teskit file has invalid URL: ${filePath} - ${url}`,
  testkitFileMissingName: (filePath: string): string =>
    `Teskit file is missing name: ${filePath}`,
};
