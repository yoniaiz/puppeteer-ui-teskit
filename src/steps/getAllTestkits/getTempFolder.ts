import { getDirName } from '../../utils/getDirName';

export const getTempFolder = () => `${getDirName(import.meta.url)}/.tmp`;
