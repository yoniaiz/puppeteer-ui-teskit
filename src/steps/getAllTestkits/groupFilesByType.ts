import path from 'path';

export const groupFilesByType = (filesPaths: string[]) => {
  const tsFiles: string[] = [];
  const jsFiles: string[] = [];
  const otherFiles: string[] = [];

  const filesMap = {
    '.ts': (filePath: string) => tsFiles.push(filePath),
    '.js': (filePath: string) => jsFiles.push(filePath),
  };

  const other = (filePath: string) => otherFiles.push(filePath);

  filesPaths.forEach((filePath) => {
    const fileExtension = path.extname(filePath);
    const handler = filesMap[fileExtension] || other;
    handler(filePath);
  });

  return {
    tsFiles,
    jsFiles,
    otherFiles,
  };
};
