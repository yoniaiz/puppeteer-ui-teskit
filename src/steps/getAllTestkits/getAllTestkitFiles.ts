import { glob } from 'glob';
import { program } from '../../utils/createProgram.js';
import { TESTKIT_FILE_EXTENSIONS } from '../../constants/testkit.constants.js';
import { logger } from '../../utils/logger.js';
import { logs } from './constants.js';

export const getAllTestkitFiles = (): string[] => {
  const folderPath = `${program.options.folder}/**/`;
  const fileName = program.options.file ? `*${program.options.file}*` : '*';
  const fileTypes = `{${TESTKIT_FILE_EXTENSIONS.join(',')}}`;

  const files = glob.sync(`${folderPath}${fileName}${fileTypes}`);

  if (!files.length) {
    logger.warn(logs.noTestkitsFound);
  }

  return files;
};
