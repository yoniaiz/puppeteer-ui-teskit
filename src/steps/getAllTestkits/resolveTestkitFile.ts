import { messages } from '../../constants/messages.js';
import { TestkitConfigFile, ResolvedTestkitFile, AxeTest } from '../../types';
import { program } from '../../utils/createProgram.js';
import { logger } from '../../utils/logger.js';
import { urlUtils } from '../../utils/url.utils.js';
import { logs } from './constants.js';

function notValidTestkit(
  testkit: TestkitConfigFile,
  filePath: string,
  baseURL?: string,
) {
  const inMainURL = !!testkit.url;

  if (!testkit) {
    logger.error(logs.testkitFileIsEmpty(filePath));
    return true;
  }

  if (
    inMainURL &&
    !urlUtils.isValidUrl(baseURL ? `${baseURL}/${testkit.url}` : testkit.url)
  ) {
    logger.error(logs.testkitFileInvalidURL(filePath, testkit.url));
    return true;
  }

  if (!testkit.name) {
    logger.error(logs.testkitFileMissingName(filePath));
  }

  const { tests } = testkit;

  if (!tests || !tests.length) {
    logger.error(messages.error.testkitFileMissingTests(filePath));
    return true;
  }

  const axeTests = tests.filter((test) => test.type === 'axe') as AxeTest[];

  for (const _test of tests) {
    if (!_test.description) {
      logger.error(
        messages.error.testkitTestMissingDescription(filePath, _test.type),
      );
      return true;
    }

    if (!inMainURL && !_test.url) {
      logger.error(
        messages.error.testkitTestMissingURL(
          filePath,
          _test.type,
          _test.description,
        ),
      );
      return true;
    }

    if (
      _test.url &&
      !urlUtils.isValidUrl(baseURL ? `${baseURL}/${_test.url}` : _test.url)
    ) {
      logger.error(
        messages.error.testkitTestInvalidURL(
          filePath,
          _test.type,
          _test.description,
          _test.url,
        ),
      );
      return true;
    }
  }

  for (const axeTest of axeTests) {
    if (!axeTest.selector) {
      logger.error(
        messages.error.testkitAxeTestMissingSelector(
          filePath,
          axeTest.description,
        ),
      );
      return true;
    }
  }

  return false;
}

export async function resolveTestkitFile(
  testkit: TestkitConfigFile,
  filePath: string,
) {
  if (testkit?.skip) {
    logger.info(messages.info.testkitFileSkipped(filePath));
    return null;
  }

  if (notValidTestkit(testkit, filePath)) {
    return null;
  }

  const { url, name, tests } = testkit;

  const testkitURL = url
    ? urlUtils.isAbsoluteURL(url)
      ? url
      : `${program.options.baseURL}/${url}`
    : '';

  const testsWithUrl = tests.map((test) => {
    if (test.url) {
      return {
        ...test,
        url: urlUtils.isAbsoluteURL(test.url)
          ? test.url
          : `${program.options.baseURL}/${test.url}`,
      };
    }
    return {
      ...test,
      url: testkitURL,
    };
  });

  const resolvedSnapFile: ResolvedTestkitFile = {
    name,
    path: filePath,
    tests: testsWithUrl,
  };

  return resolvedSnapFile;
}
