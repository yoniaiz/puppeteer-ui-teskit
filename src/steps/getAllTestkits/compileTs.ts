import tsc from 'typescript';
import { getTempFolder } from './getTempFolder';

export const compileTsSnapFiles = (filesPaths: string[]) => {
  const program = tsc.createProgram({
    rootNames: filesPaths,
    options: {
      module: tsc.ModuleKind.ESNext,
      outDir: getTempFolder(),
    },
  });

  program.emit();
};
