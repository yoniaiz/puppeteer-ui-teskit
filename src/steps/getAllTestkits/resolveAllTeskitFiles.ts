import path from 'path';
import { ResolvedTestkitFile } from '../../types/index.js';
import fs from 'fs';
import { logger } from '../../utils/logger.js';
import { logs } from './constants.js';
import { compileTsSnapFiles } from './compileTs.js';
import { groupFilesByType } from './groupFilesByType.js';
import { resolveTestkitFile } from './resolveTestkitFile.js';
import { getTempFolder } from './getTempFolder.js';

async function resolveJsSnapFile(filePath: string) {
  const testkit = await import(filePath);
  return resolveTestkitFile(testkit.default, filePath);
}

async function resolveTsSnapFile(filePath: string) {
  const fileName = path.basename(filePath).replace('.ts', '.js');
  const jsFilePath = `${getTempFolder()}/${fileName}`;

  const result = await resolveJsSnapFile(jsFilePath);

  fs.promises.unlink(jsFilePath);

  result.path = filePath;

  return result;
}

export async function resolveAllTeskitFiles(
  filesPaths: string[],
): Promise<ResolvedTestkitFile[]> {
  if (!filesPaths.length) {
    return [];
  }

  const resolvedTestkits: ResolvedTestkitFile[] = [];

  const { tsFiles, jsFiles, otherFiles } = groupFilesByType(filesPaths);

  if (tsFiles.length) {
    compileTsSnapFiles(tsFiles);

    for (const tsFilePath of tsFiles) {
      const resolvedSnapFile = await resolveTsSnapFile(tsFilePath);
      if (resolvedSnapFile) {
        resolvedTestkits.push(resolvedSnapFile);
      }
    }
  }

  for (const jsFilePath of jsFiles) {
    const resolvedSnapFile = await resolveJsSnapFile(jsFilePath);
    if (resolvedSnapFile) {
      resolvedTestkits.push(resolvedSnapFile);
    }
  }

  for (const other of otherFiles) {
    logger.warn(logs.notSupportedFileType(other));
  }

  return resolvedTestkits;
}
