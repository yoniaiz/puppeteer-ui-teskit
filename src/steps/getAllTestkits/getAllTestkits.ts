import { logger } from '../../utils/logger.js';
import { logs } from './constants.js';
import { getAllTestkitFiles } from './getAllTestkitFiles.js';
import { resolveAllTeskitFiles } from './resolveAllTeskitFiles.js';

export function getAllTestkits() {
  logger.info(logs.gettingTestkits);
  return resolveAllTeskitFiles(getAllTestkitFiles());
}
