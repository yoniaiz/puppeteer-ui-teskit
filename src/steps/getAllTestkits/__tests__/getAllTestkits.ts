import { glob } from 'glob';
import { getAllTestkits } from '../getAllTestkits.js';
import { TestkitConfigFile } from '../../../types/index.js';
import { logger } from '../../../utils/logger.js';
import { logs } from '../constants.js';

describe('getAllSnapFiles', () => {
  it('should find and return resolved files', async () => {
    const mockFilePaths = [
      'absolute/path/to/file1.js',
      'absolute/path/to/file2.js',
    ];

    const mockSnapFiles: TestkitConfigFile[] = [
      {
        name: 'file1',
        url: 'http://localhost:3000/file1',
        tests: [
          {
            type: 'axe',
            description: 'Test Axe Test',
            selector: 'body',
          },
          {
            type: 'visual',
            description: 'Test Visual Test',
          },
        ],
      },
      {
        name: 'file2',
        url: 'http://localhost:3000/file2',
        tests: [
          {
            type: 'axe',
            description: 'Test Axe Test',
            selector: 'body',
          },
          {
            type: 'visual',
            description: 'Test Visual Test',
          },
        ],
      },
    ];

    jest.mock('absolute/path/to/file1.js', () => mockSnapFiles[0], {
      virtual: true,
    });
    jest.mock('absolute/path/to/file2.js', () => mockSnapFiles[1], {
      virtual: true,
    });

    jest.spyOn(glob, 'sync').mockReturnValueOnce(mockFilePaths);

    const result = await getAllTestkits();

    expect(logger.info).toBeCalledWith(logs.gettingTestkits);
    expect(glob.sync).toHaveBeenCalledWith(
      './**/*{.ui-testkit.js,.ui-testkit.ts}',
    );
    expect(result).toMatchSnapshot();
  });
});
