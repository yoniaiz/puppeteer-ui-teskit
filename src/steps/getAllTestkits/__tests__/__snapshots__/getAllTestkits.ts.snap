// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllSnapFiles should find and return resolved files 1`] = `
[
  {
    "name": "file1",
    "path": "absolute/path/to/file1.js",
    "tests": [
      {
        "description": "Test Axe Test",
        "selector": "body",
        "type": "axe",
        "url": "http://localhost:3000/file1",
      },
      {
        "description": "Test Visual Test",
        "type": "visual",
        "url": "http://localhost:3000/file1",
      },
    ],
  },
  {
    "name": "file2",
    "path": "absolute/path/to/file2.js",
    "tests": [
      {
        "description": "Test Axe Test",
        "selector": "body",
        "type": "axe",
        "url": "http://localhost:3000/file2",
      },
      {
        "description": "Test Visual Test",
        "type": "visual",
        "url": "http://localhost:3000/file2",
      },
    ],
  },
]
`;
